processor/spanmetricsprocessor/processor.go

```diff
@@ -228,6 +228,7 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces ptrace.Traces)
 	// that should not interfere with the flow of trace data because
 	// it is an orthogonal concern to the trace flow (it should not impact
 	// upstream or downstream pipeline trace components).
+	tracesClone := traces.Clone()
```

Comment:
This to me sounds right the correct fix in this design, but this smells to me as a bad design somehow that we have to do this. The "pipelines" are correctly constructed and passed the right clone of the data by the collector (using capabilities, etc.), and somehow this breaks that assumption and considerably adds more CPU than simply executing the code on the critical path.

Reply:
Thanks for the feedback @bogdandrutu. I'm definitely open to ideas to improve the design. 😄 

The original intent with the design change was that we're effectively creating a new stream of metrics data coming out of traces and, in my mind, the propagation of trace data is an orthogonal concern to metrics. That is, the trace pipeline should not depend on, or be impacted by, other components of a downstream metrics pipeline as well as any errors and latency introduced by them. That is, I feel that spanmetricsprocessor should just be a passthrough operation for traces.

Are there existing components that translate from one telemetry type to another, whose design pattern we could follow for this processor?

> The "pipelines" are correctly constructed and passed the right clone of the data by the collector (using capabilities, etc.), and somehow this breaks that assumption and considerably adds more CPU than simply executing the code on the critical path.

I'm not very familiar with some concepts around the collector. For my learning, could you please elaborate more around what is meant by the "right clone" and "capabilities" and how this design adds considerably more CPU?

This also raises an existential question of whether this component should exist at all if it breaks some fundamental assumptions of the collector.

Reply:
I _think_ this is the only processor housed in contrib that generates telemetry of a different signal type than the signal of the pipeline the processor runs in, but it is not the only processor that "short-circuits" data immediately to an exporter (the routingprocessor also does that).  [The need to branch between signal types has been brought up before.](https://github.com/open-telemetry/opentelemetry-collector/issues/2336)

Reply:
Since this processor has already been accepted and is widely used, I don't think now is the time to discuss whether or not it should've been rejected due to the way it interacts with multiple signals.  Feels like a good discussion for the SIG meeting.

For this issue, I think cloning the traces payload is a reasonable solution to allow downstream trace processing to continue while this processor generates and forwards new metrics from its trace payload.  

If cloning the traces payload has substantial performance impact then it could be noted in this processor's README as a Warning, similar to the warning sections of the [cumulativetodelta processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/cumulativetodeltaprocessor#warnings) or [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor#warnings).  Would probably be appropriate to add a "Performance Impact" to the [Collector's standard warnings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/standard-warnings.md).

Reply:
I think I am proposing actually executing the logic on the critical path, instead of cloning to execute async the metrics logic. Not suggesting to remove the processor.

Reply:
Okay, I can do this. 

As for error handling, I'm proposing to return errors caused from any logic in this processor, but only logging errors from the metrics pipeline (i.e. `ConsumeMetrics`).

That is:
```
func (p *processorImp) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
	if err := p.tracesToMetrics(ctx, traces); err != nil {
		return err
	}

	// Forward trace data unmodified.
	return p.nextConsumer.ConsumeTraces(ctx, traces)
}

func (p *processorImp) tracesToMetrics(ctx context.Context, traces ptrace.Traces) error {
	p.lock.Lock()
	defer p.lock.Unlock()

	p.aggregateMetrics(traces)
	m, err := p.buildMetrics()
	if err != nil {
		return err
	}

	if err = p.metricsExporter.ConsumeMetrics(ctx, *m); err != nil {
		// Avoid penalising trace processing caused by metrics processing.
		p.logger.Error(err.Error())
	}

	p.resetExemplarData()
	return nil
}
```

Thoughts?

Reply:
I would do `multierr.Combine(p.tracesToMetrics(ctx, traces), p.nextConsumer.ConsumeTraces(ctx, traces))` since this processor fanout to 2 components, would call both independent.

Reply:
Good suggestion. Updated in: [9858d30](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12661/commits/9858d30b7d386d0814dd4aec7e0dfc489a0bd3a8)